microprofile/openapi/src/main/java/io/helidon/microprofile/openapi/IndexBuilder.java

```diff
@@ -0,0 +1,159 @@
+import java.util.HashSet;
+import java.util.Set;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+
+import javax.enterprise.event.Observes;
+import javax.enterprise.inject.spi.Extension;
+import javax.enterprise.inject.spi.ProcessAnnotatedType;
+
+import org.jboss.jandex.Index;
+import org.jboss.jandex.IndexReader;
+import org.jboss.jandex.IndexView;
+import org.jboss.jandex.Indexer;
+
+/**
+ * Portable extension to allow construction of a Jandex index (to pass to
+ * SmallRye OpenAPI) from CDI if no {@code META-INF/jandex.idx} file exists on
+ * the class path.
+ */
+public class IndexBuilder implements Extension {
+
+    private static final String INDEX_PATH = "/META-INF/jandex.idx";
+
+    private static final Logger LOGGER = Logger.getLogger(IndexBuilder.class.getName());
+
+    private final boolean isIndexPresentOnClasspath;
+
+    private final Set<Class<?>> annotatedTypes = new HashSet<>();
+
+    /**
+     * Creates a new instance of the index builder.
+     *
+     * @throws IOException in case of error checking for the Jandex index file
+     */
+    public IndexBuilder() throws IOException {
+        isIndexPresentOnClasspath = checkForIndexFile();
+        if (isIndexPresentOnClasspath) {
+            LOGGER.log(Level.FINE, () -> String.format("Index file %s was located and will be used", INDEX_PATH));
+        } else {
+            LOGGER.log(Level.INFO, () -> String.format(
+                    "OpenAPI support could not locate the index file %s "
+                    + "so will build an in-memory index. This slows your app start-up.%n"
+                    + "Consider using the jandex maven plug-in to build the index "
+                    + "and add it to your app at build-time",
+                    INDEX_PATH));
+        }
+    }
+
+    /**
+     * Records each type that is annotated unless a Jandex index was found on
+     * the classpath (in which case we do not need to build our own in memory).
+     *
+     * @param <X> annotated type
+     * @param event {@code ProcessAnnotatedType} event
+     */
+    <X> void processAnnotatedType(@Observes ProcessAnnotatedType<X> type) {
+        if (isIndexPresentOnClasspath) {
+            return;
+        }
+        Class<?> c = type.getAnnotatedType().getJavaClass();
+        annotatedTypes.add(c);
+    }
+
+    /**
+     * Reports an {@link IndexView} for the Jandex index that describes
+     * annotated classes for endpoints.
+     *
+     * @return {@code IndexView} describing discovered classes
+     * @throws IOException in case of error reading an existing index file or
+     * reading class bytecode from the classpath
+     */
+    public IndexView indexView() throws IOException {
+        return isIndexPresentOnClasspath ? existingIndexFileReader() : indexFromHarvestedClasses();
+    }
+
+    private IndexView existingIndexFileReader() throws IOException {
+        try (InputStream jandexIS = getClass().getResourceAsStream(INDEX_PATH)) {
```

Comment:
Consider `BufferedInputStream`.

Reply:
The `IndexReader` we pass the stream to itself wraps a `BufferedInputStream` around ours.